_posts/2019-04-25-snapshots.md

```diff
@@ -0,0 +1,135 @@
+---
+layout: post
+title: "Take <i>snapshots</i> of current screen"
+author: "beibean"
+cover: /assets/snapshots/cover.jpg
+image:
+  path: /assets/snapshots/cover.jpg
+  height: 300
+  width: 300
+---
+
+## What is this post about
+
+This post rises to explain how to take a snapshot of what LittlevGL is printing in the whole screen. In other words, press a pseudo *Ptr Scr* key.
+
+I share my way of doing it (with holder functions/variables). I am sure that this way will not fit anyone code at a first glance, but I hope that you catch the drift, so you are able to adapt it for your needs.
+
+## Why would I want to take a screen snapshot
+
+This is basically for two main reasons:
+
+- It helps to analyze what is currently being shown within the screen in a pixel level. For example, opening the image with GIMP could help us to realize that we should move elements on GUI, change colors, modify themes...
+
+- It is a must if you want to have automatic tests in your program. This could be achieved simply by comparing a reference and new output (it will be explained later).
+
+## How to obtain a binary of the whole VDB
+
+The first thing we want to achieve is to flush to a binary file the screen pixel values:
+
+- Global variables used:
+
+```c
+static FILE* pFileOut;
+static BOOL_E bSnapshot;
+```
+
+- Create a function which will be called whenever we want to take a snapshot:
+
+```c
+/** Flush buffer.
+ *
+ * Flush buffer into a binary file.
+ *
+ * @note: data pixel should be formated to uint32_t RGBA. Imagemagick requirements.
+ *
+ * @param[in] pFileName   Output binary file name.
+ *
+ */
+void DISP_IMPL_flushBuffer (const char_t* pFileName)
+{
+  pFileOut = fopen(pFileName, "w");
+
+  if (pFileOut == NULL) {
+    SIM_logMsg(("[Display] error: %s cannot be opened", pFileName));
+    return;
+  }
+
+  bSnapshot = BOOL_TRUE;
+  lv_obj_invalidate(lv_scr_act());
+  lv_refr_now();                    /* Will call our disp_drv.disp_flush function */
+  bSnapshot = BOOL_FALSE;
+
+  fclose(pFileOut);
+  SIM_logMsg(("[Display] data flushed to %s", pFileName));
+}
+```
+
+- Write every pixel received form VDB to file:
+
+```c
+void APP_DISP_lvgl_flush (const lv_color_t* pData, const DISPLAY_BOUNDS_T* const pPxBounds)
+{
+  uint32_t data;
+  if (bSnapshot == BOOL_TRUE) {
+    for (uint32_t i = 0; i < (pPxBounds->dimension.width * pPxBounds->dimension.height); i++) {
+      data = DISP_IMPL_lvgl_formatPixel(pData[i]);  /* Function for converting LittlevGL pixel format to RGB888 */
```

Comment:
`DISP_IMPL_lvgl_formatPixel` has not been implemented or explained anywhere. Could you add some information about that?